tsl/src/nodes/vector_agg/grouping_policy_batch.c

```diff
@@ -74,15 +94,15 @@ gp_batch_reset(GroupingPolicy *obj)
 
 	MemoryContextReset(policy->agg_extra_mctx);
 
-	const int naggs = list_length(policy->agg_defs);
+	const int naggs = policy->num_agg_defs;
 	for (int i = 0; i < naggs; i++)
 	{
-		VectorAggDef *agg_def = (VectorAggDef *) list_nth(policy->agg_defs, i);
-		void *agg_state = (void *) list_nth(policy->agg_states, i);
-		agg_def->func->agg_init(agg_state);
+		VectorAggDef *agg_def = &policy->agg_defs[i];
+		void *agg_state = policy->agg_states[i];
+		agg_def->func.agg_init(agg_state, 1);
```

Comment:
Is this really correct? 

agg_init() is called in a loop over `n` but most, if not all, agg_init()-functions now also loop over the agg_state array. Seems this would lead to writes outside the intended memory area, or am I missing something?

You should double-check this.

If you have the agg_state array and you know `n` I think `agg_init` can do all the initialization (including allocation of the array) internally. It seems safer than having it spread across multiple places.

Reply:
Also, should this really be agg_init(agg_state, 1) and not agg_init(agg_state, i)?

```suggestion
		agg_def->func.agg_init(agg_state, i);
```

Reply:
Ohh, I get it now. This is kind of confusing.

You are first getting the offset to the individual agg_state element, then calling agg_init() on that element and then having it run a loop internally of 1 element. This is very confusing.

Why not just pass in the vector array to agg_init() and giving the right size without having the outer loop?

Reply:
Yeah, because this is just a preparatory refactoring PR, this is only called with `n = 1`, so might look confusing. But in the hash aggregation PR, it's going to be used to initialize the aggregate states in bulk. So I pass the offset of the first state and the number of them. Is there some alternative?

Reply:
I guess I still don't understand why you need this outer loop here, end then an inner loop in the init function that you are essentially not using. 

Why not just replacing the outer loop here with a single call to `agg_def->func.agg_init(policy->agg_states, naggs)`?

Reply:
The outer loop is looping over different aggregate functions, each of them has its own state. Like `select count(*),  count(x), sum(x) from table`, you have three `VectorAggDef`s here and have to do something different for each.

Reply:
Sorry I was unclear. I understand why the outer loop exists, but I don't understand why there's an inner loop that is not used because you are only ever calling `agg_init` on the first entry in the the `agg_state` array. For example, float_sum_init(), which I assume can be called via the  `agg_init` pointer is now a loop over `n` values where n is hard-coded to 1 here instead of passing in the number of entries of the `agg_state` array.

This is why I proposed the VectorAggState struct, which would have the number of entries allocated to the `agg_state` array so that `agg_init` would know how to initialize the entire array.

But I am probably missing something. But there is something that confuses me with this code, which might be an indication of something.